feat: svm support for across (#625)
UMAprotocol/UMIPs: UMIPs/umip-179.md

#### Note
- `V3RelayData` specifies `bytes32` representation for addresses (`depositor`, `recipient`, ...) in order to interface with non-EVM chains. The EVM address supplied shall be promoted to type `bytes32` with the upper 12 bytes zeroed. 

#### SVM support

`RelayData` in SVM represents the same data as `V3RelayData` in EVM:

| Name | Type |
| :--- |:---- |
| depositor | Pubkey |
| recipient | Pubkey |
| exclusive_relayer | Pubkey |
| input_token | Pubkey |
| output_token | Pubkey |
| input_amount | u64 |

> | output_amount | u64 |

Reviewer: Maybe we've already discussed this, but if the output amount and input amounts are stored in u64, how are we going to represent amounts with 18 decimals of precision. If I'm doing my math right, we wouldn't be able to represent anything higher than ~18.4 ETH in 64 bits, for example.

Author: Indeed, this could be an issue, addressed in 770376e98e7bc934c94143304be012bd9bf48f55

| ReplacedSlowFill | 1 | The relay was initially requested to be completed via slow fill, but was subsequently fast-filled by a relayer. |
| SlowFill | 2 | The relay was completed via slow fill. |

#### SVM support

> `FillType` enum is emitted with each `FilledRelay` event on SVM with the same interpretation as when emitted with `FilledV3Relay` events on EVM.

Reviewer: Side note: the EVM event should be called `FilledRelay` as well. See https://github.com/across-protocol/contracts/blob/919bfd3f90d28a6e0013b21fc4238e8b238547f3/contracts/SpokePool.sol#L1653

Note: I think we may have the old name throughout the UMIP and some note that `FilledRelay` should be considered an equivalent event (since both are technically allowed in the current version of the UMIP).

Would be good cleanup (in a follow-up) to completely drop `FilledV3Relay` now that it's gone from the contracts.

Author: I do have a separate branch for removing `V3` from EVM as they are out of contracts now. Will post that as a separate PR once we settle on this SVM specific PR.

| Name | Type | Description |
| :--- | :--- | :---------- |
| relayer | address | The address completing relay on the destination SpokePool. |
| repaymentChainId | uint256 | The repayment chain ID requested by the relayer completing the fill. |
| relayExecutionInfo | V3RelayExecutionEventInfo | The effective `recipient`, `message` and `outputAmount`, as well as the `FillType` performed (FastFill, ReplacedSlowFill, SlowFill). |

#### Note
- Consumers of these events should append the `destinationChainId` attribute in order to avoid unintentioanlly mixing events from different chains.

#### SVM support

The `FilledRelay` event extends the `RelayData` type by applying the following adjustments:

| Name | Type | Description |
| :--- | :--- | :---------- |
| message | omitted | This field is omitted from the `FilledRelay` event in favour of the `message_hash` field. |
| message_hash | [u8; 32] | The hash of the message sent to the recipient. This is computed as `solana_program::keccak::hash(message)` where `message` is the `message` bytes in the `RelayData`. If the `message` field is empty, this will be set to `[0u8; 32]`. |

> | relayer | Pubkey | The address completing relay on the destination SpokePool, same as in EVM. |

Author: Agree, Now that relayer can specify the `repaymentAddress` that is passed to `relayer` field instead of `msg.sender` on EVM we need to fix the description both for EVM and SVM. Addressed in 6819fb72817ca839f63ba2a5061cddc0103d1e89.

### Relayer Refund Leaves
| Name | Type | Description |
| :--- | :--- | :---------- |
| amountToReturn | uint256 | Amount of `l2TokenAddress` to return to the HubPool. |
| chainId | uint256 | The SpokePool `chainId` referenced by the `RelayerRebalanceLeaf`. |
| refundAmounts | uint256[] | Ordered array of amounts of `l2TokenAddress`to be refunded to the corresponding `refundAddress`. |
| leafId | uint32 | Index of the `RelayerRefundLeaf` within the ordered array of `RelayerRefundLeaves`. |

> | l2TokenAddress | bytes32 | The SpokePool token used by this `RelayerRefundLeaf`.

Reviewer 2: ```suggestion
| l2TokenAddress | bytes32 | The SpokePool token used by this `RelayerRefundLeaf`. |
```

- Depositor refunds are to be issued as part of the relayer refund procedure.
- The `fillDeadline` timestamp shall be resolved to a block number on the destination chain in order to determine inclusion within the `Bundle Block Range`.

#### SVM support

> When evaluating if the `fillDeadline` / `fill_deadline` timestamp elapsed within the `Bundle Block Range` on the SVM destination SpokePool, one can use `getBlock` RPC method for the destination chain's bundle start and end slot and use the `blockTime` field to compare against the `fillDeadline` / `fill_deadline` timestamp.

Reviewer: What if there is no block at that slot?

Author: good point! I think the end slot must be chosen with a produced block, but if start slot happens to be empty we should use the timestamp from the last produced block before it. Added this in 585e74308e2b2c164aeb74384435929f1c9cf968

Reviewer: What does it mean if blockTime is not available? Is that just that node doesn't have the info? Or is that a hard property of the block (i.e. guaranteed to be the same on other nodes by consensus)?

Author: From [getBlockTime](https://solana.com/docs/rpc/http/getblocktime) docs:
> Each validator reports their UTC time to the ledger on a regular interval by intermittently adding a timestamp to a Vote for a particular block. A requested block's time is calculated from the stake-weighted mean of the Vote timestamps in a set of recent blocks recorded on the ledger.

So I understand `blockTime` can be null for some period of time, but not clear if that's possible once the block is finalized.

# Security considerations
Across v3 has been audited by OpenZeppelin.

> # SVM support

Reviewer 3: One structural comment - I'm not sure how to integrate this cleanly into the existing structure without creating an `SVM` section under each major section...but I'm also not sure whether it's a good fit being appended to the bottom.

Author: agree, have restructured SVM bits across relevant UMIP section

The same rules as specified in the rest of this price identifier applies in principle to Across support for SVM chains (Solana at the moment) whereas this section describes the scope of the SVM support and any considerations that are specific to SVM.

## Global Constants

As SVM does not have a concept of chain IDs it should be derived by using the 48 least significant bits of `keccak256` hashed value of the chain name. As an illustration, the chain ID for Solana mainnet is `34268394551451` that can be derived as:

```bash
cast to-dec $(cast shr $(cast shl $(cast keccak solana-mainnet) $((256-48))) $((256-48)))
```

## SVM Data Types

> Across supports only selected data types on SVM as listed in the sections below.

Reviewer 2: Nit:
```suggestion
Across supports the following data types on SVM:
```

| Name | Type | Description |
| :--- |:---- | :---------- |
| relay_data | RelayData | `RelayData` instance to uniquely associate the `SlowFill` with `FundsDeposited` and `RequestedSlowFill` events. |
| chain_id | u64 | The chain ID of the SpokePool completing the `SlowFill`. This is used only when proposing and verifying the root bundle while on execution `svm_spoke` overrides it with its configured chain ID similar as in EVM. |
| updated_output_amount | u64 | The amount sent to `recipient` as part of a `SlowFill`. |

## Events

`svm_spoke` program uses Anchor's [`emit_cpi`](https://www.anchor-lang.com/docs/features/events#emit_cpi) macro to emit events that are comparable with EVM events. On SVM Across has the following events that are relevant to this UMIP:

- FundsDeposited
- FilledRelay
- RequestedSlowFill

### FundsDeposited

The `FundsDeposited` event extends the `RelayData` type by applying the following adjustments:

| Name | Type | Description |
| :--- | :--- | :---------- |

> | origin_chain_id | omitted | This field is omitted from the `FundsDeposited` event. |

Author: I had spotted some more inconsistencies on the EVM side, but thought to better group them in a separate PR

> | origin_chain_id | omitted | This field is omitted from the `FundsDeposited` event. |

Author: as I refactored SVM bits across the UMIP I ended up also fixing any EVM inconsistencies I found

### Definitions

#### Deposits

On SVM chains a `Deposit` is defined as an instance `FundsDeposited` event.

#### Fills

On SVM chains a `Fill` is defined as an instance of `FilledRelay` event.

#### Slow Fill Requests

On SVM chains a `Slow Fill` is defined as an instance of `RequestedSlowFill` event.

#### RelayData

On SVM chains, `RelayData` is defined as an instance of the `RelayData` data type as described in the [SVM Data Types](#svm-data-types) section above.

#### Bundle Block Range

On SVM chains the `Bundle Block Range` is the pair of start and end slots for a given proposal. See [Identifying Bundle Block Ranges](#identifying-bundle-block-ranges) for guidance on identifying the `Bundle Block Range`.

#### Fill Status

A `Deposit` is considered to be `Filled` on the destination SVM chain when its `FillStatus` is resolved to `Filled` using `FillStatusAccount` as described in the [SVM Data Types](#svm-data-types) section above.

> ### Method

Reviewer 3: For the `Method` section, I am wondering if this is better integrated directly into the existing `Method` section above. In particular I think this would make sense for `Computing Relayer Repayments`, because the dance we have to do for validating `repaymentChainId` and `repaymentAddress` is already a bit complex.

- When the `Fill` `relayer` address is valid for the applied `repaymentChainId`: `relayer`, ELSE
- The `Fill` `msg.sender` address. In this case, change the applied `repaymentChainId` to the `destinationChainId` if a `PoolRebalanceRoute` exists for both the `outputToken` on the `destinationChainId` and the `inputToken` on the `originChainId` as at the bundle's hub chain end block

> If the applied `repayment address` is not valid for the applied `repaymentChainId`, the repayment shall be discarded from the relayer refund root but still considered when [computing running balances & net send amounts](#computing-running-balances--net-send-amounts). Examples of an invalid `relayer` address include:

Reviewer 3: What's the intended impact of considering the repayment in running balance & net send amount computation? Should the discarded repayment be allocated to the LP instead, or should the tokens remain in the SpokePool for some possible future reconciliation??

Author: Had discussed this with @mrice32 before that better keep the excess in the SpokePool for easier tracking and potentially refunding out of protocol via spoke admin function. Clarified this in 17736a93172e80f1bfdca44bc68e9835a86d708c
